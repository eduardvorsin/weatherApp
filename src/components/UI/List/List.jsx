import PropTypes from 'prop-types';
import classes from './List.module.css';

export default function List({
  className,
  items,
  renderInItem,
  isOrdered,
}) {
  const Tag = isOrdered ? 'ol' : 'ul';

  return (
    <Tag
      className={`${classes.list} ${className}`}
    >
      {items.map((item, index) => (
        <li
          key={typeof item === 'object' ? item.id : `${item}${index}`}
          className={classes.list__item}
        >
          {
            renderInItem
              ? renderInItem(item)
              : item
          }
        </li>
      ))}
    </Tag>
  );
}

List.defaultProps = {
  className: '',
  items: [],
  renderInItem: null,
  isOrdered: false,
};

List.propTypes = {
  className: PropTypes.string,
  renderInItem: PropTypes.func,
  isOrdered: PropTypes.bool,
  items: PropTypes.arrayOf(PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.bool,
    PropTypes.string,
    PropTypes.object,
  ])),
};
