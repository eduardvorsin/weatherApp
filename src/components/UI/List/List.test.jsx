import { render, screen } from '@testing-library/react';
import List from './List';

describe('List component tests', () => {
  it('renders correctly', () => {
    render(<List
      className="test-list"
      items={[1, 2, 3]}
      renderInItem={null}
      isOrdered
    />);

    expect(screen.getByRole('list')).toBeInTheDocument();
    expect(screen.getAllByRole('listitem')).toHaveLength(3);
  });

  it('renders correctly without props', () => {
    render(<List />);

    expect(screen.getByRole('list')).toBeInTheDocument();
    expect(screen.queryByRole('listitem')).not.toBeInTheDocument();
  });

  it('correctly renders the div passed to renderInItem', () => {
    render(<List
      items={['a', 'b', 'c', 'd', 'e']}
      renderInItem={(value) => (
        <div data-testid="test-div">{value}</div>
      )}
    />);

    expect(screen.getByRole('list')).toBeInTheDocument();
    expect(screen.getAllByRole('listitem')).toHaveLength(5);
    expect(screen.getAllByTestId('test-div')).toHaveLength(5);
    expect(screen.getAllByText(/[a-e]/i)).toHaveLength(5);
  });

  it('if the renderInItem prop is not passed, then just render <li></li> with the values', () => {
    render(<List
      className="test-list"
      items={[1, 2, 3]}
      isOrdered
    />);
    expect(screen.getByRole('list')).toBeInTheDocument();
    expect(screen.getAllByRole('listitem')).toHaveLength(3);
  });

  it('the classname is correctly assigned to the element', () => {
    render(<List
      className="test-list"
    />);

    expect(screen.getByRole('list')).toHaveClass('test-list');
  });

  it('empty list snapshot', () => {
    render(<List />);

    expect(screen.getByRole('list')).toMatchSnapshot();
  });

  it('snapshot with mock data', () => {
    render(<List
      items={['lion', 'tiger', 'cat']}
    />);

    expect(screen.getByRole('list')).toMatchSnapshot();
  });

  it('snapshot of an ordered list', () => {
    render(<List
      isOrdered
      items={['bmw', 'nissan', 'toyota']}
    />);

    expect(screen.getByRole('list')).toMatchSnapshot();
  });
});
