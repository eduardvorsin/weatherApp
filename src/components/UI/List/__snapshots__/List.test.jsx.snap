// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component tests empty list snapshot 1`] = `
<ul
  class="list "
/>
`;

exports[`List component tests snapshot of an ordered list 1`] = `
<ol
  class="list "
>
  <li
    class="list__item"
  >
    bmw
  </li>
  <li
    class="list__item"
  >
    nissan
  </li>
  <li
    class="list__item"
  >
    toyota
  </li>
</ol>
`;

exports[`List component tests snapshot with mock data 1`] = `
<ul
  class="list "
>
  <li
    class="list__item"
  >
    lion
  </li>
  <li
    class="list__item"
  >
    tiger
  </li>
  <li
    class="list__item"
  >
    cat
  </li>
</ul>
`;
